### PR TITLE
Add some return types to Array, Hash, Set and String

### PR DESCRIPTION
**Pull request by asterite**
These are generally useful, but will become more useful if https://github.com/crystal-lang/crystal/pull/11812 is merged.

### CODE REVIEW

File: src/array.cr [hunk 1]
```diff
@@ -234,7 +234,7 @@ class Array(T)
   # [2] <=> [4, 2, 3] # => -1
   # [1, 2] <=> [1, 2] # => 0
   # ```
-  def <=>(other : Array)
+  def <=>(other : Array) : Int32
     min_size = Math.min(size, other.size)
     0.upto(min_size - 1) do |i|
       n = @buffer[i] <=> other.to_unsafe[i]
```
Comment (yxhuvud) [inline src/array.cr:240]: There are cases (like for `Number`) where `<=>` is nilable, so restricting to `Int32` may be incorrect. Especially for different variants of `Float` where comparison against `NaN` gives nil back.

File: src/array.cr [hunk 2]
```diff
@@ -252,7 +252,7 @@ class Array(T)
   # ```
   #
   # See also: `#uniq`.
-  def &(other : Array(U)) forall U
+  def &(other : Array(U)) : Array(T) forall U
```
Comment (straight-shoota) [inline src/array.cr:255]: Shouldn't the type actually be `Array(T & U)`?
We currently can't express that in the type grammar. But the implementation could use this type.

File: src/set.cr [hunk 3]
```diff
@@ -203,7 +203,7 @@ struct Set(T)
   # Set{1, 1, 3, 5} & Set{1, 2, 3}               # => Set{1, 3}
   # Set{'a', 'b', 'b', 'z'} & Set{'a', 'b', 'c'} # => Set{'a', 'b'}
   # ```
-  def &(other : Set)
+  def &(other : Set) : Set(T)
```
Comment (straight-shoota) [inline src/set.cr:206]: Ditto.

File: src/array.cr (hunk 1 above)
Comment (straight-shoota) [reply src/array.cr]: We could probably still add `Int32?` as return type restriction.

File: src/array.cr (hunk 2 above)
Comment (asterite) [reply src/array.cr:255]: No, it's always Array(T). It's the elements of self that are also in the given array.

File: src/array.cr (hunk 1 above)
Comment (asterite) [reply src/array.cr]: That would be misleading if just Int32 is returned.

File: src/array.cr (hunk 2 above)
Comment (asterite) [reply src/array.cr:255]: Ah, I guess you are right. But given that we can't express that... Or how would you change it?

Comment (asterite) [reply src/array.cr:255]: Changing this would bea breaking change

Comment (yxhuvud) [reply src/array.cr:255]: No. Consider:
```
class Foo
  def ==(other)
    true
  end
end

x = [1,2,3]
y = [Foo.new]
p(x & y)
```

File: src/set.cr (hunk 3 above)
Comment (yxhuvud) [reply src/set.cr:206]: Same issue as above, though it is a little bit more complicated as the computed hash also needs to be equal.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**